### ISSUE TITLE
behaviors should be attached only after handshake is completed

### ISSUE THREAD

**Issue by noescape00** [stratisproject/StratisBitcoinFullNode#2965]
behaviors should be attached only after handshake is completed, no message should be processed except for version payload before attachment is done. 
Messages that came after the handshake but before behaviors are attached should be added to the queue and dequeued after the attachment.

**Comment by MithrilMan** [comment]
I don't agree, imo it's fine to have behaviors attached before handshake, to being allow to interact with handshake, as long as they don't process other messages before handshake

**Comment by noescape00** [comment]
Behaviors will handle in AttachCore which will be guaranteed to be called right after handshake.
So all the logic we have inside of StateChanged where we do smth if new state is handshaked- we will do those things in AttachCore. 

Or, alternatively, call StateChanged(handshake) every time after peer behaviors were attached.

**Comment by MithrilMan** [comment]
We could have two different AsyncExecutionEvent about received message
current MessageReceived could be called only if the peer state is handshaked, the other (e.g. MessageReceivedBeforeHandshake) will be called for messages when the state is connected.

Consider that actually we have cases where we look at Version payload (e.g. ProvenHeadersReservedSlotBehavior, UnreliablePeerBehavior, DropNodesBehavior, TimeSyncBehavior (this watch VerAckPayload)

I'm not worried about resources allocated for behaviors, IMO we have to refactor how behavior are instantiated and I think we should have a pool of behavior instance to be used so we can recycle them.

We could already a pool where we put behaviors detached from disconnected peers, this way before instiantiating a behavior we check if one is available from the pool.
